source/pele/xy_model_online.h

```diff
@@ -0,0 +1,232 @@
+#ifndef _PELE_XY_MODEL_ONLINE_H
+#define _PELE_XY_MODEL_ONLINE_H
+
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+#include "pele/base_potential_online.h"
+
+namespace pele {
+
+/**
+ * Adjacency list graph implementation according to S. S. Skiena,
+ * Algorithm Design Manual 
+ */
+
+struct edgenode {
+    const unsigned int y;
+    edgenode* next;
+    edgenode(const unsigned int y_, edgenode*const& next_)
+        : y(y_),
+          next(next_)
+    {}
+};
+
+class graph {
+    std::vector<edgenode*> edges;
+    std::vector<unsigned int> degree;
+    unsigned int nvertices_;
+    unsigned int nedges_;
+    const bool directed_;
+public:
+    graph(const unsigned int maxv_=1000, const bool directed__=false)
+        : directed_(directed__)
+    {
+        resize(maxv_);
+    }
+    graph(const unsigned int nvertices_input, pele::Array<size_t> x,
+        pele::Array<size_t> y)
+        : directed_(false)
+    {
+        resize(nvertices_input);
+        loadvec(nvertices_input, x, y);
+    }
+    ~graph()
+    {
+        for (unsigned int i = 0; i < nvertices_; ++i) {
+            edgenode* p = edges.at(i);
+            while (p != NULL) {
+                edgenode* tmp = p;
+                p = p->next;
+                delete tmp;
+            }
+        }
+    }
+    void loadtxt(const char* name)
+    {
+        /**
+         * File format (number of lines == nedges + 1): 
+         * First Line: nvertices and nedges
+         * Following nedges lines: two numbers:
+         * 1. vertex A, 2. vertex B
+         */
+        std::ifstream inp(name);
+        unsigned int nvertices_file;
+        unsigned int nedges_file;
+        inp >> nvertices_file;
+        inp >> nedges_file;
+        std::cout << "number of vertices in file: " << nvertices_file << "\n";
+        std::cout << "number of edges in file: " << nedges_file << "\n";
+        for (unsigned int i = 0; i < nedges_file; ++i) {
+            unsigned int x;
+            unsigned int y;
+            inp >> x;
+            inp >> y;
+            std::cout << "insert edge (" << x << ", " << y << ")\n";
+            insert_edge(x, y, directed());
+        }
+        nvertices_ = nvertices_file;
+    }
+    void loadvec(const size_t nvertices_input, pele::Array<size_t> x, pele::Array<size_t> y)
```

Comment:
also, I think nvertices_input is just there for efficiency reasons?  can we go for safety over efficiency, especially in a function that will only be called once.  You can determine nvertices by looking at the max of x and y